pgxn/neon/pagestore_smgr.c

```diff
@@ -110,6 +110,20 @@ static UnloggedBuildPhase unlogged_build_phase = UNLOGGED_BUILD_NOT_IN_PROGRESS;
 static bool neon_redo_read_buffer_filter(XLogReaderState *record, uint8 block_id);
 static bool (*old_redo_read_buffer_filter) (XLogReaderState *record, uint8 block_id) = NULL;
 
+#define MAX_LSN ((XLogRecPtr)~0)
+
+/*
+ * There are three kinds of get_page :
+ * 1. Master compute: get the latest page not older than specified LSN (horizon=Lsn::MAX)
+ * 2. RO replica: get the latest page not newer than current WAL position replica already applied (horizon=GetXLogReplayRecPtr(NULL))
+ * 3. Snapshot: get latest page not new than specified LSN (horizon=request_lsn)
+ */
+static XLogRecPtr
+neon_get_horizon(bool latest)
+{
+	return latest ? MAX_LSN : RecoveryInProgress() ? GetXLogReplayRecPtr(NULL) : InvalidXLogRecPtr; /* horizon=InvalidXlogRecPtr is replaced with request_lsn at PS */
```

Comment:
No.
InvalidXLogrecPtr=0 is low boundary, MAX_LSN=~0 is upper boundary.
This is my comment from page_service.lsn:
```
        // Horizon = 0 (INVALID) is treated as LSN interval degenerated to point [lsn,lsn].
        // It as done mostly for convenience (because such get_page commands are widely used in tests) and
        // also seems to be logical: Lsn::MAX moves upper boundary of LSN interval till last_record_lsn and
        // Lsn(0) moves upper boundary to lower boundary.
```